python/ray/data/_internal/execution/operators/actor_pool_map_operator.py

```diff
@@ -529,50 +616,61 @@ def penalty_key(actor):
             We prioritize valid actors, those with argument locality, and those that
             are not busy, in that order.
             """
-            busyness = self._num_tasks_in_flight[actor]
-            invalid = busyness >= self._max_tasks_in_flight
-            requires_remote_fetch = self._actor_locations[actor] != preferred_loc
+            busyness = self._running_actors[actor].num_tasks_in_flight
+            is_restarting = self._running_actors[actor].is_restarting
+            invalid = busyness >= self._max_tasks_in_flight or is_restarting
+            requires_remote_fetch = (
+                self._running_actors[actor].actor_location != preferred_loc
+            )
             return invalid, requires_remote_fetch, busyness
 
-        actor = min(self._num_tasks_in_flight.keys(), key=penalty_key)
-        if self._num_tasks_in_flight[actor] >= self._max_tasks_in_flight:
-            # All actors are at capacity.
+        actor = min(self._running_actors.keys(), key=penalty_key)
+        if (
+            self._running_actors[actor].num_tasks_in_flight >= self._max_tasks_in_flight
+            or self._running_actors[actor].is_restarting
+        ):
+            # All actors are at capacity or restarting.
```

Comment:
nit, not a new issue of this PR. but I think it'd be more clear if we filter the running actors by validness and then find the min.